src/main/java/com/google/devtools/build/lib/rules/cpp/SolibSymlinkAction.java

```diff
@@ -131,6 +131,37 @@ public static Artifact getDynamicLibrarySymlink(
         actionRegistry, actionConstructionContext, library, mangledName);
   }
 
+  /**
+   * Replaces shared library artifact with user specified symlink and creates related symlink
+   * action.
+   *
+   * <p>This action is performed to minimize number of -rpath entries used during linking process
+   * (by essentially "collecting" as many shared libraries as possible in the single directory),
+   * since we will be paying quadratic price for each additional entry on the -rpath.
+   *
+   * @param actionRegistry action registry of rule requesting symlink.
+   * @param actionConstructionContext action construction context of rule requesting symlink
+   * @param solibDir String giving the solib directory
+   * @param library Shared library artifact that needs to be linked.
+   * @param path Symlink path underneath the solib directory.
+   * @return linked symlink artifact.
+   */
+  public static Artifact getDynamicLibrarySymlink(
+      ActionRegistry actionRegistry,
+      ActionConstructionContext actionConstructionContext,
+      String solibDir,
+      final Artifact library,
+      PathFragment path) {
+    Preconditions.checkArgument(Link.SHARED_LIBRARY_FILETYPES.matches(library.getFilename()));
+    Preconditions.checkArgument(Link.SHARED_LIBRARY_FILETYPES.matches(path.getBaseName()));
+    Preconditions.checkArgument(!library.getRootRelativePath().getSegment(0).startsWith("_solib_"));
```

Comment:
Is this validated anywhere in the caller? (I did not see this in my visual inspection)
If any of these Precondition checks fail, it will cause a Bazel crash. We should avoid cases where bad user input can result in a Bazel crash -- we should instead throw the (checked) EvalException in such cases.

Reply:
The filetypes are already checked in the calling code. I've added a check for the `_solib_` path precondition.